### ISSUE TITLE
Bug: `isDefinitionFile` utility doesn't consider arbitrary extensions

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10911]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

eslint-plugin

### Repro Code

```TypeScript
// not applicable
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {},
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "allowArbitraryExtensions": true,
  }
}
```

### Expected Result

I expected [`isDefinitionFile`](https://github.com/typescript-eslint/typescript-eslint/blob/89f0c613e99b2419b329c1cb6277f1a286f1c98d/packages/eslint-plugin/src/util/misc.ts#L18-L29) to return `true` for file names like `.d.css.ts` (or the more general form of `{file basename}.d.{extension}.ts`) when [`allowArbitraryExtensions`](https://www.typescriptlang.org/tsconfig/#allowArbitraryExtensions) is enabled.

### Actual Result

The utility function doesn't seem to consider [`allowArbitraryExtensions`](https://www.typescriptlang.org/tsconfig/#allowArbitraryExtensions).

### Additional Info

This is a follow-up to the discussion in https://github.com/typescript-eslint/typescript-eslint/pull/10714#discussion_r1958813613.

TypeScript v5.0 added [support for `allowArbitraryExtensions`](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#allowarbitraryextensions). This seems to affect several rules that use the [`isDefinitionFile`](https://github.com/typescript-eslint/typescript-eslint/blob/89f0c613e99b2419b329c1cb6277f1a286f1c98d/packages/eslint-plugin/src/util/misc.ts#L18-L29) utility:
- `no-empty-interface`
- `no-namespace`
- `no-shadow`
- `no-unused-vars`
- `no-useless-empty-export`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.24.1` |
| `@typescript-eslint/parser`            | `8.24.1` |
| `@typescript-eslint/rule-tester`       | `8.24.1` |
| `@typescript-eslint/scope-manager`     | `8.24.1` |
| `@typescript-eslint/typescript-estree` | `8.24.1` |
| `@typescript-eslint/type-utils`        | `8.24.1` |
| `@typescript-eslint/utils`             | `8.24.1` |
| `TypeScript`                           | `5.7.3` |
| `ESLint`                               | `9.21.0` |
| `node`                                 | `22.14.0` |

**Comment by YutaK1026** [comment]
@bradzacher  Can I work on this issue? It's my first time contributing to open-source, and I'd like to try this.

**Comment by bradzacher** [comment]
https://typescript-eslint.io/contributing/issues#issue-claiming

**Comment by YutaK1026** [comment]
I see. thanks
